plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,185 @@
+--   sorted from the closest to the farthest.
+--   Gives all found paths all the way to the root directory.
+findResponsibleCabalFile :: FilePath -> IO [FilePath]
+findResponsibleCabalFile haskellFilePath = do
+  contents <- mapM (unsafeInterleaveIO . listDirectory) allDirPaths
+  let objectWithPaths = concat $ zipWith (\path content -> map (path </>) content) allDirPaths contents
+  let objectCabalExtension = filter (\c -> takeExtension c == ".cabal") objectWithPaths
+  cabalFiles <- filterM (\c -> doesFileExist c) objectCabalExtension
+  pure $ reverse cabalFiles -- sorted from closest to the haskellFilePath
+  where dirPath = dropFileName haskellFilePath
+        allDirPaths = scanl1 (</>) (splitPath dirPath)
+
+
+-- | Gives a code action that calls the command,
+--   if a suggestion for a missing dependency is found.
+--   Disabled action if no cabal files given.
+--   Conducts IO action on a cabal file to find build targets.
+hiddenPackageAction :: PluginId -> Int -> Diagnostic -> FilePath -> FilePath -> GenericPackageDescription -> IO [CodeAction]
+hiddenPackageAction plId maxCompletions diag haskellFilePath cabalFilePath gpd = do
+    buildTargets <- liftIO $ getBuildTargets gpd cabalFilePath haskellFilePath
+    case buildTargets of
+      [] -> pure $ mkCodeAction cabalFilePath Nothing <$> hiddenPackageSuggestion maxCompletions (_message diag)
+      targets -> pure $ concat [mkCodeAction cabalFilePath (Just $ buildTargetToStringRepr target) <$>
+                          hiddenPackageSuggestion maxCompletions (_message diag) | target <- targets]
+  where
+    buildTargetToStringRepr target = render $ pretty $ buildTargetComponentName target
+    mkCodeAction cabalFilePath target (suggestedDep, suggestedVersion) =
+      let
+        versionTitle = if T.null suggestedVersion then T.empty else "  version " <> suggestedVersion
+        targetTitle = case target of
+          Nothing -> T.empty
+          Just t  -> "  target " <> T.pack t
+        title = "Add dependency " <> suggestedDep <> versionTitle <> targetTitle
+
+        version = if T.null suggestedVersion then Nothing else Just suggestedVersion
+
+        params = CabalAddCommandParams {cabalPath = cabalFilePath, buildTarget = target, dependency = suggestedDep, version=version}
+        command = mkLspCommand plId (CommandId cabalAddCommand) "Execute Code Action" (Just [toJSON params])
+      in CodeAction title (Just CodeActionKind_QuickFix) (Just []) Nothing Nothing Nothing (Just command) Nothing
+
+-- | Gives a mentioned number of hidden packages given
+--   a specific error message
+hiddenPackageSuggestion :: Int -> T.Text -> [(T.Text, T.Text)]
+hiddenPackageSuggestion maxCompletions msg = take maxCompletions $ getMatch (msg =~ regex)
+  where
+    regex :: T.Text -- TODO: Support multiple packages suggestion
+    regex = "Could not load module \8216.*\8217.\nIt is a member of the hidden package [\8216\\']([a-z]+)[-]?([0-9\\.]*)[\8217\\']"
```

Comment:
Why `[\8216\\']` instead of `[\8216']`? 

`\8216.*\8217` should also account for non-Unicode quotes.

Reply:
That's an escape for the `'` symbol in this flavour or [regex](https://hackage.haskell.org/package/regex-tdfa-1.3.2.2/docs/Text-Regex-TDFA.html). Or at the very least an attempt to make one.

I would rather not try to catch all symbols in this regex, since I am looking for the package and a version separately, and add the version to the dependency if it's found.

Reply:
You don't normally need to escape quotes inside regexps, so I imagine `[\8216']` should work as well.

My point was that the module name could be `\8216.*\8217` or `'.*'`, we need to account for both possibilities.

Reply:
I think this is not about module names, but about quotes that GHC uses to surround the module name in the error message.
This particular error is generated [here](https://gitlab.haskell.org/ghc/ghc/-/blob/b57792a804ad71986238467abe6917406ffaeae5/compiler/GHC/Iface/Errors/Ppr.hs#L169) and the [quote](https://gitlab.haskell.org/ghc/ghc/-/blob/b57792a804ad71986238467abe6917406ffaeae5/compiler/GHC/Utils/Outputable.hs#L747) function used there can output both ascii quotes as well as unicode quotes.
So to make this regex robust, I think we should support both.

Reply:
Ah right, I misunderstood, thanks!